File: javascripts/bytebuilder.js
/**
 *	Builds ArrayBuffers of data in an easy to use way.
 *
 *	@class
 */
function ByteBuilder() {
	this.data = "";
	this.currentByte = 0;
	this.position = 0;
}

/**
 *	Adds a single bit to the buffer. The basis of the other write methods.
 *
 *	@param {boolean} bit - Bit to be written to the buffer.
 */
ByteBuilder.prototype.appendBit = function(bit) {
	if (bit) {
		this.currentByte |= (1 << this.position);
	}
	this.position++;

	if (this.position >= 8) {
		this.position = 0;
		
		this.data += String.fromCharCode(this.currentByte);

		this.currentByte = 0;
	}
}

/**
 *	Appends multiple bits from an integer, up to a specified number of bits
 *
 *	@param {number} bits - The source number for the appended bits.
 *	@param {number} length - Number of bits from the number to append.
 */
ByteBuilder.prototype.appendBits = function(bits, length) {
	for (var i=0; i<length; i++) {
		this.appendBit((bits & (1 << i)) > 0);
	}
}

/**
 *	Appends multiple bits from an integer, but the resulting bit order is reversed. Added for the sake of Exp-Golomb encoding and endian problems.
 *
 *	@param {number} bits - The source number for the appended bits.
 *	@param {number} length - Number of bits from the number to append.
 */
ByteBuilder.prototype.appendBitsReversed = function(bits, length) {
	for (var i=0; i<length; i++) {
		this.appendBit((bits & (1 << ((length-1)-i))) > 0);
	}
}

/**
 *	Appends 8 bytes from the input to the buffer
 *
 *	@param {number} byte - The number from which 8 bits will be appended.
 */
ByteBuilder.prototype.appendByte = function(byte) {
	this.appendBits(byte, 8);
}

/**
 *	Appends 32 bytes from the input to the buffer
 *
 *	@param {number} num - The number from which 32 bits will be appended.
 */
ByteBuilder.prototype.appendInt32 = function(num) {
	this.appendBits(num, 32);
}

/**
 *  Appends a number to the buffer using exponential-Golomb encoding
 *
 *  @param {number} num - The number to encode
 */
ByteBuilder.prototype.appendExpGolomb = function(num) {
    num++;

    var bits = Math.floor(Math.log(num)/Math.LN2);

    for (var i=0; i<bits; i++) {
        this.appendBit(false);
    }

    this.appendBitsReversed(num, bits+1);
}

/**
 *	Returns an ArrayBuffer containing the data from the builder. 
 *
 *	@param {ArrayBuffer} [bufferToExtend] - Optional. If specified, the data from this will be at prepended to the builder's data.
 *
 *	@returns {ArrayBuffer} ArrayBuffer representing data collected by the ByteBuilder
 */
ByteBuilder.prototype.toBuffer = function(bufferToExtend) {
	var length = this.position > 0 ? this.data.length + 1 : this.data.length;
	var offset = 0;
	
	var extending = false;
	
	if (typeof(bufferToExtend) !== "undefined" && bufferToExtend != null) {
		length += bufferToExtend.byteLength;
		offset = bufferToExtend.byteLength;
		extending = true;
	}
	
	var list = new Uint8Array(length);
	
	if (extending) {
		var view = new Uint8Array(bufferToExtend);
		for (var i=0; i<view.length; i++) {
			list[i] = view[i];
		}
	}
	
	for (var i=0; i<this.data.length; i++) {
		list[i+offset] = this.data.charCodeAt(i);
	}
	if (this.position > 0) {
		list[this.data.length + offset] = this.currentByte;
	}
	
	return list.buffer;
}

//##################################################################################################################
//##################################################################################################################
//##################################################################################################################

/**
 *	Reads data from an ArrayBuffer in an easy to use way.
 *
 *	@class
 *	@param {ArrayBuffer} bytes - The source ArrayBuffer to read from.
 *	@param {number} [offset=0] - Optional. Byte offset to start reading from.
 */
function ByteReader(bytes, offset) {
	var o = 0;
	if (typeof(offset) !== "undefined") {
		o = offset;
	}

	this.bytes = new Uint8Array(bytes, o);
	this.position = 0;
}

/**
 *	Reads a single bit at a specified position. For internal use only.
 *
 *	@param {number} position - Bit offset to read from.
 *
 *	@returns {boolean} State of the bit at position.
 */
ByteReader.prototype.read = function(position) {
	var bytepos = Math.floor(position / 8);
	var bitpos = position % 8;

	if (bytepos >= this.bytes.length) {
	    throw "Attempted to read out of range: "+bytepos;
	}

	var byte = this.bytes[bytepos];
	
	return (byte & (1 << bitpos)) > 0;
}

/**
 *	Reads one bit from the buffer.
 *
 *	@returns {boolean} State of the next bit in the buffer.
 */
ByteReader.prototype.readBit = function() {
	var val = this.read(this.position);
	this.position++;
	return val;
}

/**
 *	Reads multiple bits from the buffer.
 *
 *	@param {number} bitcount - How many bits to read.
 *
 *	@returns {number} A number representing the next bitcount bits from the buffer.
 */
ByteReader.prototype.readBits = function(bitcount) {
	if (bitcount > 32) {
		throw "bitcount may not exceed 32";
	}
	var val = 0;
	
	for (var i=0; i<bitcount; i++) {
		if (this.readBit()) {
			val |= (1 << i);
		}
	}
	
	return val;
}

/**
 *	Reads multiple bits from the buffer, but in reverse order. Mostly added for Exp-Golomb encoding and endian issues.
 *
 *	@param {number} bitcount - How many bits to read.
 *
 *	@returns {number} A number representing the next bitcount bits from the buffer, read in reverse order.
 */
ByteReader.prototype.readBitsReversed = function(bitcount) {
		if (bitcount > 32) {
		throw "bitcount may not exceed 32";
	}
	var val = 0;
	
	for (var i=0; i<bitcount; i++) {
		if (this.readBit()) {
			val |= (1 << ((bitcount-1)-i));
		}
	}
	
	return val;
}

/**
 *	Reads 8 bits from the buffer.
 *
 *	@returns {number} A number representing the next 8 bits from the buffer.
 */
ByteReader.prototype.readByte = function() {
	return this.readBits(8);
}

/**
 *	Reads 32 bits from the buffer.
 *
 *	@returns {number} A number representing the next 32 bits from the buffer.
 */
ByteReader.prototype.readInt32 = function() {
	return this.readBits(32);
}

/**
 *  Reads a number encoded using exponential-Golomb encoding.
 */
ByteReader.prototype.readExpGolomb = function() {
    var bits = 0;

    while (true) {
        if (this.readBit()) {
            this.position--;
            break;
        } else {
            bits++;
        }
    }

    return (this.readBitsReversed(bits+1))-1;
}

File: web/scripts/includes/bytebuilder.dart
import 'dart:typed_data';
import 'dart:math';

/// Builds a compacted [ByteBuffer] of data with syntax similar to [StringBuffer].
class ByteBuilder {
	/// Internal buffer.
	StringBuffer _data = new StringBuffer();
	/// Current working byte, bits are appended to this up to 8, then it is added to the buffer.
	int _currentbyte = 0;
	/// Bit position within the current working byte.
	int _position = 0;

	/// Creates a new ByteBuilder with an empty buffer.
	ByteBuilder();

	/// Appends a single bit to the buffer.
	void appendBit(bool bit) {
		if (bit) {
			_currentbyte |= (1 << _position);
		}
		_position++;
		if (_position >= 8) {
			_position = 0;
			_data.writeCharCode(_currentbyte);
			_currentbyte = 0;
		}
	}

	/// Appends [length] bits of [bits] to the buffer.
	void appendBits(int bits, int length) {
		for (int i=0; i<length; i++) {
			appendBit(bits & (1 << i) > 0);
		}
	}

	/// Appends [length] bits of [bits] to the buffer, in reverse order.
	///
	/// Used by [appendExpGolomb].
	void appendBitsReversed(int bits, int length) {
		for (int i=0; i<length; i++) {
			appendBit(bits & (1 << ((length-1)-i)) > 0);
		}
	}

	/// Appends 8 bits of [byte] to the buffer.
	void appendByte(int byte) {
		appendBits(byte, 8);
	}

	/// Appends 16 bits of [i] to the buffer.
	void appendShort(int i) {
		appendBits(i, 16);
	}

	/// Appends 32 bits of [i] to the buffer.
	void appendInt32(int i) {
		appendBits(i, 32);
	}

	/// Appends [i] to the buffer using Exponential-Golomb encoding.
	///
	/// [Wikipedia reference](https://en.wikipedia.org/wiki/Exponential-Golomb_coding)
	void appendExpGolomb(int i) {
		i++;

		int bits = log(i)~/LN2;

		for (int i=0; i<bits; i++) {
			this.appendBit(false);
		}

		this.appendBitsReversed(i, bits+1);
	}

	/// Appends all numbers in [bits] to the buffer as [length] bit long segments.
	void appendAllBits(List<int> bits, int length) {
		for (int number in bits) {
			this.appendBits(number, length);
		}
	}

	/// Appends all numbers in [bytes] as bytes.
	void appendAllBytes(List<int> bytes) {
		this.appendAllBits(bytes, 8);
	}

	/// Appends all numbers in [numbers] using Exponential-Golomb encoding.
	void appendAllExpGolomb(List<int> numbers) {
		for (int number in numbers) {
			this.appendExpGolomb(number);
		}
	}

	/// Creates a new [ByteBuffer] containing the data in this ByteBuilder.
	ByteBuffer toBuffer([ByteBuffer toExtend = null]) {
		int length = _position > 0 ? _data.length+1 : _data.length;
		int offset = 0;

		////print(this._data.toString());

		if (toExtend != null) {
			length += toExtend.lengthInBytes;
			offset = toExtend.lengthInBytes;
		}

		Uint8List list = new Uint8List(length);

		if (toExtend != null) {
			Uint8List view = new Uint8List.view(toExtend);
			for (int i=0; i<view.length; i++) {
				list[i] = view[i];
			}
		}

		String data = _data.toString();

		for (int i=0; i<data.length; i++) {
			list[i+offset] = data.codeUnitAt(i);
		}
		if (_position > 0) {
			list[data.length+offset] = _currentbyte;
		}

		return list.buffer;
	}

	/// Convenience function for pretty-printing a [ByteBuffer].
	static void prettyPrintByteBuffer(ByteBuffer buffer) {
		Uint8List list = new Uint8List.view(buffer);

		StringBuffer sb = new StringBuffer("Bytes: ${buffer.lengthInBytes} [");

		for (int i=0; i<list.length; i++) {
			sb.write("0x${list[i].toRadixString(16).padLeft(2,"0").toUpperCase()}");
			if (i < list.length-1) {
				sb.write(", ");
			}
		}

		sb.write("]");

		print(sb.toString());
	}
}

/// Reads a [ByteBuffer] as a stream of bits.
class ByteReader {
	/// Source buffer.
	ByteData _bytes;
	/// Current bit position within the buffer.
	int _position = 0;

	/// Creates a new ByteReader reading from [bytes]. The start position will be offset by [offset] bytes.
	ByteReader(ByteBuffer bytes, [int offset = 0]) {
		this._bytes = bytes.asByteData(offset);
	}

	/// Internal method for reading a bit at a specific position. Use read for getting single bits from the buffer instead.
	bool _read(int position) {
		int bytepos = (position / 8.0).floor();
		int bitpos = position % 8;

		int byte = _bytes.getUint8(bytepos);

		return byte & (1 << bitpos) > 0;
	}

	/// Reads the next bit from the buffer.
	bool readBit() {
		bool val = this._read(this._position);
		_position++;
		return val;
	}

	/// Reads the next [bitcount] bits from the buffer.
	int readBits(int bitcount) {
		if (bitcount > 32) {
			throw new ArgumentError.value(bitcount,"bitcount may not exceed 32");
		}
		int val = 0;

		for (int i=0; i<bitcount; i++) {
			if (readBit()) {
				val |= (1 << i);
			}
		}

		return val;
	}

	/// Reads the next [bitcount] bits from the buffer, in reverse order.
	///
	/// Used by readExpGolomb.
	int readBitsReversed(int bitcount) {
		if (bitcount > 32) {
			throw new ArgumentError.value(bitcount,"bitcount may not exceed 32");
		}
		int val = 0;

		for (int i=0; i<bitcount; i++) {
			if (readBit()) {
				val |= (1 << ((bitcount-1)-i));
			}
		}

		return val;
	}

	/// Reads the next 8 bits from the buffer.
	int readByte() {
		return readBits(8);
	}

	/// Reads the next 16 bits from the buffer.
	int readShort() {
		return readBits(16);
	}

	/// Reads the next 32 bits from the buffer.
	int readInt32() {
		return readBits(32);
	}

	/// Reads a number encoded with Exponential-Golomb encoding from the buffer.
	///
	/// The bit length read depends upon the encoded number.
	///
	/// [Wikipedia reference](https://en.wikipedia.org/wiki/Exponential-Golomb_coding)
	int readExpGolomb() {
		int bits = 0;

		while (true) {
			if (this.readBit()) {
				this._position--;
				break;
			} else {
				bits++;
			}
		}

		return (this.readBitsReversed(bits+1))-1;
	}
}